### PR TITLE
OTLP receiver: Convert also metric metadata

### PR DESCRIPTION
**Pull request by aknuds1**
Improve OTLP receiver, so it also converts metric metadata. @gouthamve's [closed PR](https://github.com/prometheus/prometheus/pull/13439) made me realize we had left this out.

### CODE REVIEW

File: storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go [hunk 1]
```diff
@@ -37,7 +37,6 @@ type Settings struct {
 	DisableTargetInfo         bool
 	ExportCreatedMetric       bool
 	AddMetricSuffixes         bool
-	SendMetadata              bool
```
Comment (aknuds1) [inline storage/remote/otlptranslator/prometheusremotewrite/metrics_to_prw.go:40]: Not used.

File: storage/remote/otlptranslator/prometheusremotewrite/otlp_to_openmetrics_metadata.go [hunk 2]
```diff
@@ -42,36 +41,3 @@ func otelMetricTypeToPromMetricType(otelMetric pmetric.Metric) prompb.MetricMeta
 	}
 	return prompb.MetricMetadata_UNKNOWN
 }
-
-func OtelMetricsToMetadata(md pmetric.Metrics, addMetricSuffixes, allowUTF8 bool) []*prompb.MetricMetadata {
```
Comment (aknuds1) [inline storage/remote/otlptranslator/prometheusremotewrite/otlp_to_openmetrics_metadata.go:46]: Not in use.

**Review by yeya24 — APPROVED** [review]
LGTM

File: storage/remote/otlptranslator/prometheusremotewrite/otlp_to_openmetrics_metadata.go (hunk 2 above)
Comment (aknuds1) [reply storage/remote/otlptranslator/prometheusremotewrite/otlp_to_openmetrics_metadata.go:46]: Converting metadata in `FromMetrics` is clearly more:

* Performant -> you don't need an extra loop over the OTLP data, you don't need to call `prometheustranslator.BuildCompliantName` twice, you don't need to make an extra slice to convert from `[]*prompb.MetricMetadata` to `[]prompb.MetricMetadata`.
* Readable/maintainable -> you need far less code by integrating into `FromMetrics`

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,7 @@
 * [ENHANCEMENT] Scraping, rules: handle targets reappearing, or rules moving group, when out-of-order is enabled. #14710
 * [ENHANCEMENT] Tools: add debug printouts to promtool rules unit testing #15196
 * [ENHANCEMENT] Scraping: support Created-Timestamp feature on native histograms. #14694
+* [ENHANCEMENT] OTLP receiver: Convert also metric metadata. #15416
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply CHANGELOG.md:14]: Done.

File: storage/remote/write_handler.go [hunk 4]
```diff
@@ -526,6 +526,7 @@ func (h *otlpWriteHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 	err = h.rwHandler.write(r.Context(), &prompb.WriteRequest{
 		Timeseries: converter.TimeSeries(),
+		Metadata:   converter.Metadata(),
```
[the remark this answers is not in the document]
Comment (aknuds1) [reply storage/remote/write_handler.go:529]: We already lack tests for converted time series, right? I think adding another suite of tests is out of scope for this PR, as it only adds metadata in addition to time series. The existing test suite is already updated.

**Review by ArthurSens — APPROVED** [review]

**Comment by bboreham** [comment]
Just so I understand: there was a function `OtelMetricsToMetadata`, but it was never called.
You replaced that function with similar code inside other functions, which are called, so we now get metadata ?

**Comment by aknuds1** [comment]
> You replaced that function with similar code inside other functions, which are called, so we now get metadata ?

@bboreham correct. You can compare to @gouthamve's [PR](https://github.com/prometheus/prometheus/pull/13439) where he was going to use `OtelMetricsToMetadata` for the same end result, but he closed it (the PR).

**Pull request merged**